482p3/man/man8/btrace.8
.TH BTRACE 8
.SH NAME
btrace \- block-level tracing interface
.SH SYNOPSIS
\fBbtrace\fR \fBstart\fR \fIdevice\fR \fIentries\fR
.PP
\fBbtrace\fR \fBstop\fR \fIdevice\fR \fIfile\fR
.PP
\fBbtrace\fR \fBreset\fR \fIdevice\fR
.PP
\fBbtrace\fR \fBdump\fR \fIfile\fR
.SH DESCRIPTION
The \fBbtrace\fR tool is the user interface to MINIX3's block-level tracing
facility. It allows one to start, stop, and reset tracing, and dump a trace
output file in a somewhat human-readable format.
.SH COMMANDS
.TP 10
\fBstart\fR
This command starts tracing all block requests on the given \fIdevice\fR. Each
block request takes up one entry in the log. The \fIentries\fR parameter
specifies the allocation side of the log in the driver process, in number of
(32-byte) entries. Once the log is full, no more entries will be added.
.TP 10
\fBstop\fR
This command stops tracing on the given \fIdevice\fR, and dumps the resulting
log to the given output \fIfile\fR.
.TP 10
\fBreset\fR
This command stops tracing on the given \fIdevice\fR and resets the state of
the block tracer. This should be useful only in emergency situations.
.TP 10
\fBdump\fR
Dump the contents of a log file generated earlier with \fBbtrace stop\fR, in
human-readable format. Heavy users of the block tracing facility will probably
want to write their own tools for parsing and visualizing dump files.
.SH LIMITATIONS
Only one block device can be traced per driver at once. It is therefore also
not possible to trace a device and all its partitions at the same time. The
tracing facility has been designed for tracing activity of a single file
system, in which case these limitations are not important.
.PP
The log will always start with a \fIclose\fR operation on the device, since
\fBbtrace\fR closes the file descriptor used to instruct the driver to start
tracing. Similarly, for logs that have not already filled up during tracing,
the last entry will be a \fBbtrace\fR-triggered \fIopen\fR operation.
.SH EXAMPLES
.TP 35
.B btrace start /dev/c2d0 10240
# Start a block trace on c2d0.
.TP 35
.B btrace stop /dev/c2d0 outfile
# Stop the block trace on c2d0.
.TP 35
.B btrace dump outfile
# View the output of the trace.
.SH "SEE ALSO"
.BR ioctl (2).
.SH AUTHOR
David van Moolenbroek <david@minix3.org>


482p3/commands/btrace/btrace.c
/* Block trace command line tool */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <minix/btrace.h>
#include <minix/u64.h>
#include <sys/ioc_block.h>

static btrace_entry buf[BTBUF_SIZE];

static void usage(char *name)
{
  printf("usage:\n"
    "%s start <device> <nr_entries>\n"
    "%s stop <device> <file>\n"
    "%s reset <device>\n"
    "%s dump <file>\n",
    name, name, name, name);

  exit(EXIT_FAILURE);
}

static void btrace_start(char *device, int nr_entries)
{
  int r, ctl, devfd;
  size_t size;

  if ((devfd = open(device, O_RDONLY)) < 0) {
	perror("device open");
	exit(EXIT_FAILURE);
  }

  size = nr_entries;
  if ((r = ioctl(devfd, BIOCTRACEBUF, &size)) < 0) {
	perror("ioctl(BIOCTRACEBUF)");
	exit(EXIT_FAILURE);
  }

  ctl = BTCTL_START;
  if ((r = ioctl(devfd, BIOCTRACECTL, &ctl)) < 0) {
	perror("ioctl(BIOCTRACECTL)");

	size = 0;
	(void) ioctl(devfd, BIOCTRACEBUF, &size);

	exit(EXIT_FAILURE);
  }

  close(devfd);
}

static void btrace_stop(char *device, char *file)
{
  int r, ctl, devfd, outfd;
  size_t size;

  if ((devfd = open(device, O_RDONLY)) < 0) {
	perror("device open");
	exit(EXIT_FAILURE);
  }

  if ((outfd = open(file, O_CREAT|O_TRUNC|O_WRONLY, 0600)) < 0) {
	perror("file open");
	exit(EXIT_FAILURE);
  }

  ctl = BTCTL_STOP;
  if ((r = ioctl(devfd, BIOCTRACECTL, &ctl)) < 0) {
	perror("ioctl(BIOCTRACECTL)");
	exit(EXIT_FAILURE);
  }

  for (;;) {
	if ((r = ioctl(devfd, BIOCTRACEGET, buf)) < 0) {
		perror("ioctl(BIOCTRACEGET)");
		break;
	}

	if (r == 0) break;

	size = r * sizeof(buf[0]);
	if ((r = write(outfd, (char *) buf, size)) != size) {
		if (r < 0) perror("write");
		else fputs("short write\n", stderr);
	}
  }

  close(outfd);

  size = 0;
  if ((r = ioctl(devfd, BIOCTRACEBUF, &size)) < 0) {
	perror("ioctl(BIOCTRACEBUF)");
	exit(EXIT_FAILURE);
  }

  close(devfd);
}

static void btrace_reset(char *device)
{
  size_t size;
  int r, ctl, devfd;

  if ((devfd = open(device, O_RDONLY)) < 0) {
	perror("device open");
	exit(EXIT_FAILURE);
  }

  ctl = BTCTL_STOP;
  (void) ioctl(devfd, BIOCTRACECTL, &ctl);

  size = 0;
  if ((r = ioctl(devfd, BIOCTRACEBUF, &size)) < 0) {
	perror("ioctl(BIOCTRACEBUF)");
	exit(EXIT_FAILURE);
  }

  close(devfd);
}

static void dump_entry(btrace_entry *entry)
{
  switch (entry->request) {
  case BTREQ_OPEN: printf("OPEN"); break;
  case BTREQ_CLOSE: printf("CLOSE"); break;
  case BTREQ_READ: printf("READ"); break;
  case BTREQ_WRITE: printf("WRITE"); break;
  case BTREQ_GATHER: printf("GATHER"); break;
  case BTREQ_SCATTER: printf("SCATTER"); break;
  case BTREQ_IOCTL: printf("IOCTL"); break;
  }

  printf(" request\n");

  switch (entry->request) {
  case BTREQ_OPEN:
	printf("- access:\t%x\n", entry->size);
	break;
  case BTREQ_READ:
  case BTREQ_WRITE:
  case BTREQ_GATHER:
  case BTREQ_SCATTER:
	printf("- position:\t%08lx%08lx\n",
		ex64hi(entry->position), ex64lo(entry->position));
	printf("- size:\t\t%u\n", entry->size);
	printf("- flags:\t%x\n", entry->flags);
	break;
  case BTREQ_IOCTL:
	printf("- request:\t%08x\n", entry->size);
	break;
  }

  printf("- start:\t%u us\n", entry->start_time);
  printf("- finish:\t%u us\n", entry->finish_time);
  if (entry->result == BTRES_INPROGRESS)
	printf("- result:\t(in progress)\n");
  else
	printf("- result:\t%d\n", entry->result);
  printf("\n");
}

static void btrace_dump(char *file)
{
  int i, r, infd;

  if ((infd = open(file, O_RDONLY)) < 0) {
	perror("open");
	exit(EXIT_FAILURE);
  }

  for (;;) {
	if ((r = read(infd, (char *) buf, sizeof(buf))) <= 0)
		break;

	r /= sizeof(buf[0]);

	for (i = 0; i < r; i++)
		dump_entry(&buf[i]);
  }

  if (r < 0) perror("read");

  close(infd);
}

int main(int argc, char **argv)
{
  int num;
  char *name = argv[0];

  if (argc < 3) usage(name);

  if (!strcmp(argv[1], "start")) {
	if (argc < 4) usage(name);

	num = atoi(argv[3]);

	if (num <= 0) usage(name);

	btrace_start(argv[2], num);
  }
  else if (!strcmp(argv[1], "stop")) {
	if (argc < 4) usage(name);

	btrace_stop(argv[2], argv[3]);
  }
  else if (!strcmp(argv[1], "reset")) {
	btrace_reset(argv[2]);
  }
  else if (!strcmp(argv[1], "dump")) {
	btrace_dump(argv[2]);
  }
  else usage(name);

  return EXIT_SUCCESS;
}
